### DISCUSSION TITLE
When I use a cuda function, how do I find the call chain corresponding to the function in the driver.

### DISCUSSION THREAD

**Discussion by wenhaopang** [NVIDIA/open-gpu-kernel-modules#337] [Q&A] [answered]
When I use a cuda function, how do I find the call chain corresponding to the function in the driver.

**Comment by aritger** [comment, marked as answer]
There isn't a clear mapping between function calls made to the CUDA user-mode driver and functions within the open kernel modules
in this repository.  There is a lot of logic implemented in the CUDA user-mode driver.  So, it isn't a thin wrapper over kernel system calls (in contrast to something like glibc).

To answer your question, probably the best thing to do would be to trace the entry points into the kernel drivers (e.g., spinkling some printks or similar).  Most of the interesting calls will start with the fileops in kernel-open/nvidia/nv.c, and most of those will be ioctl.  I would suggest to start with this call chain:

    kernel-open/nvidia/nv.c:nvidia_ioctl()
      src/nvidia/arch/nvalloc/unix/src/osapi.c:rm_ioctl()
        src/nvidia/arch/nvalloc/unix/src/escape.c:RmIoctl()

**Reply by wenhaopang** [reply]
Thank you so much for your suggestion.

**Reply by YusufKhan-gamedev** [reply]
You can also see this commit by Dave Airlie if you want to see how a reference compute implementation would work:
https://gitlab.freedesktop.org/nouveau/mesa/-/commit/dce9441c56d1b189746696d9c0dc2c7b09844ccb

**Reply by wenhaopang** [reply]
Thank you very much.

**Reply by YusufKhan-gamedev** [reply]
The nvk vulkan driver also uses headers from uvm(the cl****.h ones) and iirc CUDA requires uvm so perhaps also look there(especially the pushbuffer part of uvm).

**Reply by wenhaopang** [reply]
OK, thanks.

**Comment by J-StrawHat** [comment]
I am reaching out to kindly inquire if there have been any new progress or findings. Have you been able to uncover any information regarding the call chain for CUDA functions within the driver? Any insights or details you could share would be immensely helpful and greatly appreciated.
